brave/brave-ios
# Fix #6820: Allow to permanently delete Sync accounts

## Summary of Changes

This Pull Request includes the "Delete Sync Chain" functionality on iOS which is the equivalent of the Permanently delete sync account on Android and desktop https://github.com/brave/brave-core/pull/13562.

The brave-core PR exposes https://github.com/brave/brave-core/pull/17369 the interface for necessary functionality on brave sync service such as 

- Adds support for iOS Chromium Sync API delete chain functionality so we can deleted sync chains and detect deleted accounts.
- Adds observer methods regarding to deleted account functionality so we can update our data and UI on iOS side.
- Adds support for Sync Profile so we can give proper errors while joining deleted accounts.

The PR consists of 3 major parts 

1. The Delete Action

The delete button is added to Sync Settings and this is handling functionality like the actual delete chain action, removing sync observers, resetting the chain and clearing the local preferences. The delete action will show an alert to user similar to Android but using the design used for pop-ups in iOS settings.

2. The Account Deleted Error

After validating sync code words, a new callback is added to validate if account is not deleted and if it is deleted according to error type, app shows account deleted error alert and like delete chain action it will removing sync observers and clears the local preferences.

3. The Account Deleted Observer

Adding an observer and handling different cases for different parts of the application when the sync chain is deleted from another device.

Security Ticket: https://github.com/brave/security/issues/1207


This pull request fixes #6820

## Submitter Checklist:


## Test Plan - Screenshots::


### Test Plan 1: Delete Sync Chain

- Create a Sync Chain either from Desktop/Android or from iOS and join the same chain with the devices
- Press button Delete Sync Account and approve deletion
- Observe iOS device will navigate back to settings
- Observe desktop will update and show join sync chain screen.


https://user-images.githubusercontent.com/6643505/221986792-c76e73cf-9dbb-4bd9-8f07-253ae8f0b343.mov

![221986853-338580c7-45d1-4dc9-b403-dced60f07809](https://user-images.githubusercontent.com/6643505/222449090-3df315e0-9c7c-471e-a883-11cdeb57caac.png)



### Test Plan 2: Join a Deleted Sync Chain

- Create a Sync Chain either from Desktop/Android and save the code word or qr code
- Try to join the chain with that code words - qr code
- Observe iOS will return an error about deleted chain


https://user-images.githubusercontent.com/6643505/221990656-13257ba4-a063-4249-a869-fd1063e20a00.mov

![221986853-338580c7-45d1-4dc9-b403-dce809111122](https://user-images.githubusercontent.com/6643505/222502380-d429af59-4446-4391-803b-5d0636476994.png)


### Test Plan 3: Receive a notification from deleted sync chain

- Create a Sync Chain either from Desktop/Android or from iOS and join the same chain with the devices
- Press button Delete Sync Account and approve deletion on Desktop or Android (delete sync account using other device pair)
- Observe iOS device will navigate back to settings


https://user-images.githubusercontent.com/6643505/222004522-29dd2d7e-2067-48c7-bc54-7d390bf05a83.mov


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
